pkg/patterns/declarative/metrics.go

```diff
@@ -0,0 +1,342 @@
+	if err != nil {
+		return false, err
+	}
+
+	if mapping.Scope.Name() == meta.RESTScopeNameNamespace {
+		return true, nil
+	} else {
+		return false, nil
+	}
+}
+
+type reconcileCounter struct {
+	groupVersionKind string
+	namespaced       bool
+	counterVec       *prometheus.CounterVec
+}
+
+func reconcileCounterFor(gvk schema.GroupVersionKind, scope meta.RESTScopeName) reconcileCounter {
+	if scope == meta.RESTScopeNameNamespace {
+		return reconcileCounter{groupVersionKind: gvkString(gvk), namespaced: true, counterVec: reconcileCount}
+	} else {
+		return reconcileCounter{groupVersionKind: gvkString(gvk), namespaced: false, counterVec: reconcileCount}
+	}
+}
+
+func (rc reconcileCounter) reconcileWith(request reconcile.Request) {
+	if rc.namespaced {
+		if len(request.Namespace) == 0 {
+			request.Namespace = DefaultNamespace
+		}
+	}
+	rc.counterVec.WithLabelValues(rc.groupVersionKind, request.Namespace, request.Name).Inc()
+}
+
+type reconcileFailureCounter struct {
+	groupVersionKind string
+	namespaced       bool
+	counterVec       *prometheus.CounterVec
+}
+
+func reconcileFailureCounterFor(gvk schema.GroupVersionKind, scope meta.RESTScopeName) reconcileFailureCounter {
+	if scope == meta.RESTScopeNameNamespace {
+		return reconcileFailureCounter{groupVersionKind: gvkString(gvk), namespaced: true, counterVec: reconcileFailure}
+	} else {
+		return reconcileFailureCounter{groupVersionKind: gvkString(gvk), namespaced: false, counterVec: reconcileFailure}
+	}
+}
+
+func (rfc reconcileFailureCounter) failureWith(request reconcile.Request) {
+	if rfc.namespaced {
+		if len(request.Namespace) == 0 {
+			request.Namespace = DefaultNamespace
+		}
+	}
+	rfc.counterVec.WithLabelValues(rfc.groupVersionKind, request.Namespace, request.Name).Inc()
+}
+
+type objectRecorder struct {
+	groupVersionKind string
+	gaugeVec         *prometheus.GaugeVec
+}
+
+func objectRecorderFor(gvk schema.GroupVersionKind) objectRecorder {
+	return objectRecorder{groupVersionKind: gvkString(gvk), gaugeVec: managedObjectsRecord}
+}
+
+func (or objectRecorder) Set(namespace, name string, num float64) {
+	if namespace == ClusterScoped {
+		or.gaugeVec.WithLabelValues(or.groupVersionKind, "", name).Set(num)
+	} else {
+		or.gaugeVec.WithLabelValues(or.groupVersionKind, namespace, name).Set(num)
+	}
+}
+
+var globalObjectTracker *objectTracker = newObjectTracker()
+
+type objectTracker struct {
+	mu         sync.Mutex
+	mgr        manager.Manager
+	trackedGVK map[schema.GroupVersionKind]*gvkTracker
+}
+
+// Call this before kubectl.Apply
+func (ot *objectTracker) addIfNotPresent(object *manifest.Object) error {
+	ot.mu.Lock()
+	defer ot.mu.Unlock()
+
+	gvk, ns, name := object.GroupVersionKind(),
+		object.UnstructuredObject().GetNamespace(),
+		object.UnstructuredObject().GetName()
+
+	namespaced, err := namespaced(ot.mgr, gvk)
+	if err != nil {
+		return err
+	}
+
+	if namespaced {
+		if len(ns) == 0 {
+			ns = DefaultNamespace
+		}
+	} else {
+		ns = ClusterScoped
```

Comment:
Does prometheus allow the empty value?  I suspect that would be more intuitive if so.

Reply:
Promtheus allows empty value. However, here, I need something to distinguish namespaced object and cluster-scoped object. 
I'll set metrics namespace label to empty value if namespace is not specified because it isn't reasonable to say empty namespace is `default` namespace as you point out in another comment.